util/collections/machine_filters.go

```diff
@@ -158,6 +158,41 @@ func HasUnhealthyCondition(machine *clusterv1.Machine) bool {
 	return conditions.IsFalse(machine, clusterv1.MachineHealthCheckSucceededCondition) && conditions.IsFalse(machine, clusterv1.MachineOwnerRemediatedCondition)
 }
 
+// HasMissingNodeOrUnhealthyControlPlaneComponents returns a filter to find all unhealthy control plane machines that
+// does not have a Kubernetes node or have any of the following control plane component conditions set to False:
+// APIServerPodHealthy, ControllerManagerPodHealthy, SchedulerPodHealthy, EtcdPodHealthy & EtcdMemberHealthy (if using managed etcd).
+// It is different from the HasUnhealthyCondition func which checks MachineHealthCheck conditions.
+func HasMissingNodeOrUnhealthyControlPlaneComponents(isEtcdManaged bool) Func {
+	controlPlaneMachineHealthConditions := []clusterv1.ConditionType{
+		controlplanev1.MachineAPIServerPodHealthyCondition,
+		controlplanev1.MachineControllerManagerPodHealthyCondition,
+		controlplanev1.MachineSchedulerPodHealthyCondition,
+	}
+	if isEtcdManaged {
+		controlPlaneMachineHealthConditions = append(controlPlaneMachineHealthConditions,
+			controlplanev1.MachineEtcdPodHealthyCondition,
+			controlplanev1.MachineEtcdMemberHealthyCondition,
+		)
+	}
+	return func(machine *clusterv1.Machine) bool {
+		if machine == nil {
+			return false
+		}
+		if !HasNode()(machine) {
```

Comment:
For this first iteration I would drop this from the filter, because there is no way to determine if a node is still provisioning or failed to provision, and I think we should avoid to intentionally delete a control plane machine still provisioning (aka let's avoid to interrupt when a new etcd member is joining if possible).

Without this check the PR should still be ok to address wrong configuration of static pods (api server, etcd, etc), but it doesn't help in case of wrong configuration for kubelet.
This use case can be addressed by MHC with nodeStartup timeout

Reply:
Thanks @fabriziopandini 
1) MHC is not turned on by default. So the wrong configuration for kubelet can not be covered when MHC is not enabled.
2) This filter is called by selectMachineForScaleDown(). In this case, do we still need to care if a node is still provisioning or failed to provision?
```
func selectMachineForScaleDown(ctx context.Context, controlPlane *internal.ControlPlane, outdatedMachines collections.Machines) (*clusterv1.Machine, error) {
	machines := controlPlane.Machines
	switch {
	case controlPlane.MachineWithDeleteAnnotation(outdatedMachines).Len() > 0:
		machines = controlPlane.MachineWithDeleteAnnotation(outdatedMachines)
	case controlPlane.MachineWithDeleteAnnotation(machines).Len() > 0:
		machines = controlPlane.MachineWithDeleteAnnotation(machines)
+	case controlPlane.UnhealthyMachinesWithMissingNodeOrUnhealthyControlPlaneComponents(outdatedMachines).Len() > 0:
+		machines = controlPlane.UnhealthyMachinesWithMissingNodeOrUnhealthyControlPlaneComponents(outdatedMachines)
	case outdatedMachines.Len() > 0:
		machines = outdatedMachines
	}
```

Reply:
The problem is that when a node is in provisioning and we then delete it we don't know if the etcd member is already joined or not. So the question is in which state we leave the cluster then

Reply:
> MHC is not turned on by default. So the wrong configuration for kubelet can not be covered when MHC is not enabled.

I think with the current way this works there will be always cases that KCP doesn't cover if MHC is disabled. The only way out of that would be to re-implement MHC within KCP.

Reply:
Make sense. Thank you both! Updated in the [lastest commit](https://github.com/kubernetes-sigs/cluster-api/pull/10196/commits/adcd88f350332f0cb3067ac69c219631ce8cb605), with a comment to explain why not using !HasNode().